`ChannelManager` documentation refresh

jkczyz opened the pull request:
`ChannelManager`'s documentation could use a little love. It needs:
- a clear explanation of its purpose and use
- how it can be parameterized
- how it relates to other components (e.g. `ChainMonitor` / `ChannelMonitors`)
- operational requirements
- example usage

All this needs to be organized in a way that is approachable, discoverable, and doesn't overwhelm the user. Some documentation may need to be consolidated or placed at a higher-level so it is not missed.

This PR re-writes `ChannelManager`'s summary, adds sections to delineate the existing docs, and proposes new sections. It still has a ways to go. One consideration is how to split the documentation between `ChannelManager`, its methods, the higher-level `channelmanager` module docs (yet to be re-written), and even higher-level modules. Or if the module layout should be flattened a bit to foster better discovery.

Feedback encouraged!

ConorOkus commented:
Concept ACK.

It looks good so far, code snippets inline will be very helpful IMO.

jkczyz commented:
Latest push fills in the "`ChannelManager` vs `ChannelMonitor`" and "Initialization" sections, the latter of which now has a code example. Considering whether the docs on `ChannelManagerReadArgs` should be moved there and if the example should go into more depth regarding feeding chain data.

As mentioned earlier, we may want to move portions of this to module-level docs instead of having it all at the struct-level.

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -951,11 +951,121 @@ where
 ///
-/// Implements [`ChannelMessageHandler`], handling the multi-channel parts and passing things through
-/// to individual Channels.
+/// [`ChannelManager`] is parameterized by a number of components, which in aggregate facilitate
```
Yeah, this goes back to how we want to breakdown the organization of these docs across modules and structs as mentioned in the PR description. The draft has everything on `ChannelManager`, but my feeling is we should minimize those docs and move the sectioned breakdown to higher-level docs, likely on a module akin to how `Iterator` is documented in the `std` library.

So I can see moving everything after this list to the module-level, linking to it from here. I'm fine with also putting a short mention of `ChannelMonitor` here and linking to the particular module-level section doc that goes into detail.

ConorOkus replied:
I do think the number of foundational components needed to get a `ChannelManager` up a running is worth noting. I've found it does help devs with their mental models and how things connect together

dunxen replied:
Yip. Agree with @ConorOkus on knowing what it makes use of upfront. At least in my experience it helps to know what it _is_ (including what it's composed of) first and then more deeply go into what it _does_ and _how_, then all the other caveats to take note of.

jkczyz replied:
@dunxen Any opinion on whether most of the docs should move to the module-level as seems to be common in Rust standard library docs?

dunxen replied:
Might expand the module docs quite a bit. Otherwise either seems fine if that’s not a concern (or if it’s bare now anyway)

yellowred replied:
For me as a dev, It would be helpful to expand the first sentence. 
Firstly, why is it 2 things: state machine and payment management? I always refer to it to just state machine, but then maybe I am wrong here.
Secondly, as a state machine it would be helpful to know the possible states and actions that change the state, where is the state being held, saved, loaded?
Lastly, what logic does this state machine implement? What logic goes into the parameters and what is in the state machine and why.

jkczyz replied:
Thanks for the feedback! I'm not sure exactly how to address these yet, but wanted to respond in case it helps you or others see a way to improve these docs. We'll likely land this as-is but I'm definitely open to improvements in follow-ups. Contributions are more than welcome, too.

> For me as a dev, It would be helpful to expand the first sentence. Firstly, why is it 2 things: state machine and payment management? I always refer to it to just state machine, but then maybe I am wrong here.

While there may be some refactoring that would allow us to have a cleaner separation of channel state machine from payment management, they are intrinsically linked as payment-related messages cause channel state transitions. Some of which may result in payment completion, failure, retry, or timeout.

> Secondly, as a state machine it would be helpful to know the possible states and actions that change the state, where is the state being held, saved, loaded?

That seems more like internal developer docs. The goal of `ChannelManager` is to abstract away the details of the state machine from developers. That said, there are probably higher-level states exposed in `ChannelDetails` and `RecentPaymentDetails` that could be useful to better document there. I believe we are at least linking to those docs here albeit indirectly through methods that return those types.

> Lastly, what logic does this state machine implement? What logic goes into the parameters and what is in the state machine and why.

The Lightning BOLT specification. :) Or at least some of it. Kidding aside, I think we want to err on the side of brevity here, but documentation on the linked parameterized types are in many instances ripe for going into further details, if they haven't already. I'll be the first to admit our docs need some work. Hopefully this is just a first step.

jkczyz replied:
> > Secondly, as a state machine it would be helpful to know the possible states and actions that change the state, where is the state being held, saved, loaded?
> 
> That seems more like internal developer docs. The goal of `ChannelManager` is to abstract away the details of the state machine from developers. That said, there are probably higher-level states exposed in `ChannelDetails` and `RecentPaymentDetails` that could be useful to better document there. I believe we are at least linking to those docs here albeit indirectly through methods that return those types.
> 

A further point here is how such transitions result in `Event` generation. We do go into that in the examples, but yes those could possibly use some higher-level docs as it relates to state transitions.

jkczyz replied:
> That said, I do agree the first sentence could be a bit more user-relevant, and maybe get a second sentence in the first paragraph. Maybe the "Channelmanager...facilitate sending, forwarding, and receiving payments through lightning channels" should move to the first sentence, and the second paragraph intro can just be "ChannelManager is parameterized"?

Ah, ok, that sounds good. See latest push.

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -951,11 +951,121 @@ where
+///
+/// An implementation of [`chain::Watch`], such as [`ChainMonitor`], is responsible for aggregating
+/// these [`ChannelMonitor`]s and applying any [`ChannelMonitorUpdate`]s to them. It then monitors
+/// for any pertinent on-chain activity, broadcasting claims as needed.
```
Much better, thanks!

jkczyz reviewed (commented):
Filled in more of the docs. Just need to do the part on payments and decide on the organization.

ConorOkus commented:
Is there a way to generate preview links for these at all?

jkczyz commented:
> Is there a way to generate preview links for these at all?

Not that I'm immediately aware of or can see from a glance through https://github.com/marketplace.

Locally, you can run `cargo doc -p lightning --open`.

jkczyz commented:
Latest push includes docs for `ChannelManager`'s BOLT 12 API.

jkczyz commented on lightning/src/ln/channelmanager.rs:
````diff
@@ -1068,11 +1068,610 @@ where
+/// # }
+/// ```
+///
+/// TODO: Consider re-writing ChannelManagerReadArgs docs and moving here.
````
Sure, take a stab at it!

jkczyz replied:
Could you wrap at 100 characters so it's easier to read?

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -1068,11 +1068,610 @@ where
+/// for any pertinent on-chain activity, enforcing claims as needed.
+///
+/// This division of off-chain management and on-chain enforcement is what allows for interesting
+/// node setups. See the [`chain`] module for further details.
```
Re-phrased a bit with a little more succinct wording of what you suggested.

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -1108,19 +1717,53 @@ where
 ///
+/// [`ChainMonitor`]: crate::chain::chainmonitor::ChainMonitor
+/// [`MessageHandler`]: crate::ln::peer_handler::MessageHandler
+/// [`OnionMessenger`]: crate::onion_message::OnionMessenger
```
Note, this will need to be updated once rebased since the re-export was removed.

dunxen commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -1068,11 +1068,616 @@ where
+/// use bitcoin::BlockHash;
+/// use bitcoin::network::constants::Network;
+/// use lightning::chain::BestBlock;
+/// # use lightning::chain::channelmonitor::ChannelMonitor;
+/// use lightning::ln::channelmanager::{ChainParameters, ChannelManager, ChannelManagerReadArgs};
+/// # use lightning::routing::gossip::NetworkGraph;
+/// use lightning::util::config::UserConfig;
+/// use lightning::util::ser::ReadableArgs;
+///
+/// # fn read_channel_monitors() -> Vec<ChannelMonitor<lightning::sign::InMemorySigner>> { vec![] }
+/// # fn example<
+/// #     'a,
+/// #     L: lightning::util::logger::Logger,
+/// #     S: for <'b> lightning::routing::scoring::LockableScore<'b, ScoreLookUp = SL>,
+/// #     SL: lightning::routing::scoring::ScoreLookUp<ScoreParams = SP>,
+/// #     SP: Sized,
+/// #     R: lightning::io::Read,
+/// # >(
+/// #     fee_estimator: &dyn lightning::chain::chaininterface::FeeEstimator,
+/// #     chain_monitor: &dyn lightning::chain::Watch<lightning::sign::InMemorySigner>,
+/// #     tx_broadcaster: &dyn lightning::chain::chaininterface::BroadcasterInterface,
+/// #     router: &lightning::routing::router::DefaultRouter<&NetworkGraph<&'a L>, &'a L, &S, SP, SL>,
+/// #     logger: &L,
+/// #     entropy_source: &dyn lightning::sign::EntropySource,
+/// #     node_signer: &dyn lightning::sign::NodeSigner,
+/// #     signer_provider: &dyn lightning::sign::SignerProvider<EcdsaSigner = lightning::sign::InMemorySigner>,
+/// #     best_block: lightning::chain::BestBlock,
+/// #     current_timestamp: u32,
+/// #     mut reader: R,
+/// # ) -> Result<(), lightning::ln::msgs::DecodeError> {
```
Silly question, but why is the example fn signature commented out in these docs?

jkczyz replied:
Commenting it out is primarily to make the documentation more concise. You can argue it's hard to tell what types those identifiers are referring to. But hopefully the naming and cross-referencing the documentation for functions where they are used is sufficient. That way the docs can focus more on demonstrating the particular concept / functionality.

The general use a function here is as a way to avoid needing to have a lot of concrete types for type parameterizations in the example itself. It also means the code is not actually execute, only compiled since we aren't calling the function anywhere.

jkczyz replied:
Also, not sure if it was clear but "commenting out" with `#` simply hides those lines from the generated docs.

dunxen reviewed (commented):
Awesome! New (and existing) users and contributors are going to love this.

dunxen commented:
CI just needs attention but LGTM

jkczyz commented:
> Feel free to squash with some clarification for [#2704 (comment)](https://github.com/lightningdevkit/rust-lightning/pull/2704#discussion_r1537919203) and I think we can basically land this.

Squashed and rebased to fix the doc linking. Regarding the linked comment, happy to find some better wording if you prefer. I was tying to keep it brief in the first sentence, while the second sentences sorta links channels and payments by saying "... sending, forwarding, and receiving payments _through_ lightning channels".

jkczyz commented:
> Ugh, looks like doctests are now failing on `cfg=c_bindings`

Should be fixed

dunxen reviewed (approved):
Set of changes since last ACK LGTM.
